scene/3d/collision_object_3d.cpp

```diff
@@ -207,26 +226,40 @@ void CollisionObject3D::_clear_debug_shapes() {
 		ShapeData::ShapeBase *shapes = shapedata.shapes.ptrw();
 		for (int i = 0; i < shapedata.shapes.size(); i++) {
 			ShapeData::ShapeBase &s = shapes[i];
-			if (s.debug_shape) {
-				s.debug_shape->queue_delete();
-				s.debug_shape = nullptr;
-				--debug_shape_count;
+			if (s.debug_shape.is_valid()) {
+				RS::get_singleton()->free(s.debug_shape);
+				s.debug_shape = RID();
+				if (s.shape.is_valid() && s.shape->is_connected("changed", callable_mp(this, &CollisionObject3D::_update_shape_data))) {
+					s.shape->disconnect("changed", callable_mp(this, &CollisionObject3D::_update_shape_data));
+				}
 			}
 		}
 	}
-
-	debug_shape_count = 0;
+	debug_shapes_count = 0;
 }
 
 void CollisionObject3D::_update_shape_data(uint32_t p_owner) {
-	if (is_inside_tree() && get_tree()->is_debugging_collisions_hint() && !Engine::get_singleton()->is_editor_hint()) {
+	if (_are_collision_shapes_visible()) {
 		if (debug_shapes_to_update.is_empty()) {
-			call_deferred("_update_debug_shapes");
+			callable_mp(this, &CollisionObject3D::_update_debug_shapes).call_deferred({}, 0);
 		}
 		debug_shapes_to_update.insert(p_owner);
 	}
 }
 
+void CollisionObject3D::_on_transform_changed() {
+	if (debug_shapes_count > 0 && !debug_shape_old_transform.is_equal_approx(get_global_transform())) {
+		debug_shape_old_transform = get_global_transform();
+		for (Map<uint32_t, ShapeData>::Element *E = shapes.front(); E; E = E->next()) {
+			ShapeData &shapedata = E->get();
+			const ShapeData::ShapeBase *shapes = shapedata.shapes.ptr();
+			for (int i = 0; i < shapedata.shapes.size(); i++) {
+				RS::get_singleton()->instance_set_transform(shapes[i].debug_shape, shapedata.xform * debug_shape_old_transform);
```

Comment:
The transforms are not multiplied in the same order as in `_update_debug_shapes`, so one of them is incorrect. It would be nice to check cases with transforms on shapes to make sure the transform is done the right way.

Reply:
The `_update_debug_shapes` one is correct.
> It would be nice to check cases with transforms on shapes to make sure the transform is done the right way.

I have several tests written during the development of #45783 and this PR (including an enhanced version of what you gave me in #45783).
I can clean up then a bit and upload them here if you want.

Reply:
> I have several tests written during the development of #45783 and this PR (including an enhanced version of what you gave me in #45783).
I can clean up then a bit and upload them here if you want.

If it's not too much extra work, that would be great! Especially if you have some tests with custom collision shapes.

Reply:
Here is a bunch of tests (it's not a project) [CollisionObject_debug_meshes.zip](https://github.com/godotengine/godot/files/6420042/CollisionObject_debug_meshes.zip)